pkg/controller/servicebinding/sbrcontroller.go

```diff
@@ -106,34 +108,49 @@ func isOfKind(obj runtime.Object, kind string) bool {
 	return strings.EqualFold(obj.GetObjectKind().GroupVersionKind().Kind, kind)
 }
 
+// check whether annotations specify bindings
+func isAnnotationValid(ann map[string]string) bool {
+	for k := range ann {
+		if strings.HasPrefix(k, annotations.ServiceBindingOperatorAnnotationPrefix) {
+			return true
+		}
+	}
+	return false
+}
+
 // updateEvent returns a predicate handler function.
 func updateFunc(logger *log.Log) func(updateEvent event.UpdateEvent) bool {
 	return func(e event.UpdateEvent) bool {
 		isSecret := isOfKind(e.ObjectNew, "Secret")
 		isConfigMap := isOfKind(e.ObjectNew, "ConfigMap")
 
 		if isSecret || isConfigMap {
-			dataFieldsAreEqual, err := compareObjectFields(e.ObjectOld, e.ObjectNew, "data")
+			dataFieldsAreEqual, err := compareRuntimeObjectFields(e.ObjectOld, e.ObjectNew, "data")
 			if err != nil {
 				logger.Error(err, "error comparing object fields: %s", err.Error())
 				return false
 			}
 			logger.Debug("Predicate evaluated for Secret/ConfigMap", "dataFieldsAreEqual", dataFieldsAreEqual)
 			return !dataFieldsAreEqual
 		}
-		specsAreEqual, err := compareObjectFields(e.ObjectOld, e.ObjectNew, "spec")
+		specsAreEqual, err := compareRuntimeObjectFields(e.ObjectOld, e.ObjectNew, "spec")
 		if err != nil {
 			logger.Error(err, "error comparing object's spec fields: %s", err.Error())
 			return false
 		}
-		statusAreEqual, err := compareObjectFields(e.ObjectOld, e.ObjectNew, "status")
+		statusAreEqual, err := compareRuntimeObjectFields(e.ObjectOld, e.ObjectNew, "status")
 		if err != nil {
 			logger.Error(err, "error comparing object's status fields: %s", err.Error())
 			return false
 		}
-		shouldReconcile := !specsAreEqual || !statusAreEqual
-		logger.Debug("Resource update event received", "GVK", e.ObjectNew.GetObjectKind(), "spec changed", !specsAreEqual, "status changed", !statusAreEqual, "should reconcile", shouldReconcile)
-
+		annotationsAreEqual := true
+		annOld := e.MetaOld.GetAnnotations()
+		annNew := e.MetaNew.GetAnnotations()
+		if isAnnotationValid(annOld) && isAnnotationValid(annNew) {
+			annotationsAreEqual = reflect.DeepEqual(annOld, annNew)
+		}
```

Comment:
Could you elaborate what are you trying to achieve in here? 

My interpretation is that annotations are deeply compared if and only if both annotation maps have at least one valid service binding annotation; if this is indeed true, then both annotation maps are deeply compared and, if any change is identified by `reflect.DeepEqual`, the resource will be reconciled.

I expected to read here something along the following lines:

```go
annotationsAreEqual, err := compareAnnotations(e.MetaOld.GetAnnotations(), e.MetaNew.GetAnnotations())
if err != nil { ... }
```
`compareAnnotations` could then perform, such as `compareRuntimeObjectFields` the appropriate comparison for annotations and return an error in situations the comparison could not be performed.

Additionally, having a broader API surface with reasonable tests increase the understanding the reasons behind a failed comparison.

Reply:
runtime.Object is being used as parameters for `compareRuntimeObjectFields`  and we need meta.Object. Hence I didn't use that function.
I can have a compareAnnotations function performing the reflect.DeepEqual. Would that suffice?